File: templates/py/app/views/bar_aggregated.py
# This block is used to aggregate the data from the Bar table into a materialized view
from moose_lib.dmv2 import MaterializedView, MaterializedViewOptions
from pydantic import BaseModel

class BarAggregated(BaseModel):
    day_of_month: int
    total_rows: int
    rows_with_text: int
    total_text_length: int
    max_text_length: int

# The query to create the materialized view, which is executed when the block is set up
select_query = """
SELECT
  toDayOfMonth(utc_timestamp) as day_of_month,
  count(primary_key) as total_rows,
  countIf(has_text) as rows_with_text,
  sum(text_length) as total_text_length,
  max(text_length) as max_text_length
FROM bar
GROUP BY toDayOfMonth(utc_timestamp)
"""

barAggregatedMV = MaterializedView[BarAggregated](MaterializedViewOptions(
    select_statement=select_query,
    table_name="bar_aggregated",
    materialized_view_name="bar_aggregated_mv",
    order_by_fields=["day_of_month"]
))

File: templates/typescript/app/views/barAggregated.ts
import typia from "typia";
import { MaterializedView, sql } from "@514labs/moose-lib";
import { BarPipeline } from "../datamodels/models";

interface BarAggregated {
  dayOfMonth: number & typia.tags.Type<"int64">;
  totalRows: number & typia.tags.Type<"int64">;
  rowsWithText: number & typia.tags.Type<"int64">;
  totalTextLength: number & typia.tags.Type<"int64">;
  maxTextLength: number & typia.tags.Type<"int64">;
}

const barTable = BarPipeline.table!;
export const BarAggregatedMV = new MaterializedView<BarAggregated>({
  tableName: "BarAggregated",
  materializedViewName: "BarAggregated_MV",
  orderByFields: ["dayOfMonth"],
  selectStatement: sql`SELECT
    toDayOfMonth(${barTable.columns.utcTimestamp}) as dayOfMonth,
    count(${barTable.columns.primaryKey}) as totalRows,
    countIf(${barTable.columns.hasText}) as rowsWithText,
    sum(${barTable.columns.textLength}) as totalTextLength,
    max(${barTable.columns.textLength}) as maxTextLength
  FROM ${barTable}
  GROUP BY toDayOfMonth(utcTimestamp)
  `,
});
